### ISSUE TITLE
RuntimeError: Models must be owned by only a single document, CDSView(id='1142', ...) is already in a doc

### ISSUE THREAD

**Issue by PezAmaury** [PatrikHlobil/Pandas-Bokeh#44]
Hi!

I have a problem that is likely due to interaction with Bokeh, but it occurs only when using pandas_bokeh plots and not Bokeh plots directly.

When trying to make a dashboard with pandas_bokeh with layout or plot_grid, I get the following error:
```
Traceback (most recent call last):
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/app/Influenza/src/main/python/main.py", line 370, in on_click_check
    bokehPlots.bokehPlots(self.data_demo, self.data_case)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/app/Influenza/src/main/python/bokehPlots.py", line 162, in bokehPlots
    pandas_bokeh.plot_grid([[p_noso]])
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/pandas_bokeh/base.py", line 53, in plot_grid
    show(layout)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/pandas_bokeh/base.py", line 134, in show
    bokeh.plotting.show(obj, browser, new, notebook_handle, notebook_url)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/io/showing.py", line 145, in show
    return _show_with_state(obj, state, browser, new, notebook_handle=notebook_handle)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/io/showing.py", line 183, in _show_with_state
    _show_file_with_state(obj, state, new, controller)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/io/showing.py", line 166, in _show_file_with_state
    filename = save(obj, state=state)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/io/saving.py", line 86, in save
    _save_helper(obj, filename, resources, title, template)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/io/saving.py", line 148, in _save_helper
    html = file_html(obj, resources, title=title, template=template)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/embed/standalone.py", line 297, in file_html
    with OutputDocumentFor(models, apply_theme=theme, always_new=_always_new) as doc:
  File "/usr/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/embed/util.py", line 138, in OutputDocumentFor
    doc.add_root(model)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/document/document.py", line 304, in add_root
    self._pop_all_models_freeze()
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/document/document.py", line 1019, in _pop_all_models_freeze
    self._recompute_all_models()
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/document/document.py", line 1042, in _recompute_all_models
    a._attach_document(self)
  File "/mnt/c/Users/thiabaud/Documents/IOGH_Influenza/DataAnalysis/Scripts/Python/pyqt5_venv/lib/python3.6/site-packages/bokeh/model.py", line 714, in _attach_document
    raise RuntimeError("Models must be owned by only a single document, %r is already in a doc" % (self))
RuntimeError: Models must be owned by only a single document, CDSView(id='1142', ...) is already in a doc

```
When looking for a solution, Bokeh's documentation and overall Stackoverflow answers say that it's likely to be either an object being used several time (which it isn't), or the source dataframe must not be used several time - but I tried using different dataframes for each plot without any luck. Creating a layout or using plot_grid result in the same error.

I'm running bokeh 1.3.4 and pandas_bokeh 0.3
As I previously mentioned, the error occurs only when I use plots created with pandas_bokeh, the bokeh alone plot is properly dashboarding on its own with plot_grid.

**Comment by PatrikHlobil** [comment]
Hi @PezAmaury ,

Could you give a minimal example where this error occurs. Maybe this is related to issue #31 .

Best 

Patrik

**Comment by PezAmaury** [comment]
Hi @PatrikHlobil 

On the following script, if I use plot_grid with figure p, I have no problems (it's pure bokeh). If I use any of the others (p_types, p_origin, etc...) I get the error:

```
#!/usr/bin/env python
# coding: utf-8

# ### Imports
import os
import pandas as pd
import numpy as np
from datetime import datetime
import math
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_colwidth', -1)
# ### Displays all outputs
#from IPython.core.interactiveshell import InteractiveShell
#InteractiveShell.ast_node_interactivity = "all"
# ### Bokeh imports
from bokeh.io import show, output_notebook, output_file
from bokeh.plotting import figure
from bokeh.models import ColumnDataSource
from bokeh.palettes import Spectral6
from bokeh.models import FactorRange, Legend
from bokeh.palettes import OrRd6, GnBu6, RdPu6
from bokeh.palettes import OrRd3, GnBu3, RdPu3
#output_notebook()
# ### Pandas_bokeh
import pandas_bokeh
#pandas_bokeh.output_notebook()
# ### Custom imports
import generateWeeklyData


def bokehPlots(data_demo, data_case):
    
    colors = ["#e84d60", "#718dbf", "#c9d9d3"]
    dirName = './html/'
    if not os.path.exists(dirName):
        os.mkdir(dirName)
    start_date = datetime(2018,11,1)
    end_date = datetime(2019,4,30)
    # ## Loads data: both the demographics & the episode data
    #data_demo = pd.read_csv('data_demo.csv')
    #data_case = pd.read_csv('data_case.csv')

    # ### Generate weekly data
    pyramid_data = generateWeeklyData.count_year_gender(data_demo).reset_index()
    pyramid_data.columns = ['centre_id','age_group','Men','Women','other']
    pyramid_data['centre_id'] = 'Centre_#' + pyramid_data['centre_id'].astype(str)

    data_service =  generateWeeklyData.count_casesType(data_case, start_date, end_date)
    data_service = data_service.reset_index().drop(columns = ['other'])
    data_service.columns = ['week','centre_id','Nosocomial','Comm. acquired','N/A','Type A','Type B','Type N/A','Deaths','No deaths','Medicine',
                            'Geriatrics','ICU','Surgery','Paediatrics','Other']
    data_service['centre_id'] = 'Centre_#' + data_service['centre_id'].astype(str)
    data_service['All cases'] = data_service['Nosocomial']+data_service['Comm. acquired']+data_service['N/A']

    data_noso = generateWeeklyData.count_noso(data_case, start_date, end_date).reset_index()
    data_noso.columns = ['week','centre_id','Type A','Type B','Type N/A','Deaths','No deaths','Medicine',
                        'Geriatrics','ICU','Surgery','Paediatrics','Other']
    data_noso['centre_id'] = 'Centre_#' + data_noso['centre_id'].astype(str)
    
    all_centres = data_service['centre_id'].unique()
    
    # ### Creates data for the plots
    df = data_service[['week','centre_id','All cases']]
    table = pd.pivot_table(df, values='All cases', index='week',
                            columns=['centre_id'], aggfunc=np.sum)
    
    all_pyramid = pyramid_data.groupby(['age_group','centre_id']).sum().drop(columns = ['other']).unstack(level=0)
    all_pyramid.columns = all_pyramid.columns.swaplevel(0, 1)
    all_pyramid.sort_index(axis=1, level=0, inplace=True)
    all_pyramid = all_pyramid.reset_index()
    all_pyramid = all_pyramid.set_index('centre_id')
    
    # ALL CENTRES
    
    if len(all_centres) > 1:

        #pandas_bokeh.output_file(dirName + "ALL_cases.html")
        p_cases = table.plot_bokeh.bar(figsize=(800, 600),
            xlabel="Week",
            ylabel="Number of new cases",
            title="Total number of cases",
            legend="top_right",
            stacked=True,
            alpha=0.6, panning = False, zooming = False,
            vertical_xlabel=True, colormap = GnBu6)
        
        #pandas_bokeh.output_file(dirName + "All_origin.html")
        p_origin = data_service.groupby('week').sum()[['Nosocomial','Comm. acquired','N/A']].plot_bokeh.bar(figsize=(800, 600),
             xlabel="Week",
             ylabel="Number of new cases",
             title="Overall distribution of infection origin",
             legend="top_right",
             stacked=False,
             alpha=0.6, panning = False, zooming = False,
             vertical_xlabel=True, colormap = colors)
        
        #pandas_bokeh.output_file(dirName + "ALL_types.html")
        p_types = data_service.groupby('week').sum()[['Type A','Type B','Type N/A']].plot_bokeh.bar(figsize=(800, 600),
             xlabel="Week",
             ylabel="Number of new cases",
             title="Overall distribution of infection type",
             legend="top_right",
             stacked=True,
             alpha=0.6, panning = False, zooming = False,
             vertical_xlabel=True, colormap = colors)
        
        #pandas_bokeh.output_file(dirName + "ALL_service.html")
        p_service = data_service.groupby('week').sum()[['Medicine','Geriatrics','ICU','Surgery','Paediatrics','Other']].plot_bokeh.bar(figsize=(800, 600),
             xlabel="Week",
             ylabel="Number of new cases",
             title="Overall distribution of cases in units",
             legend="top_right",
             stacked=True,
             alpha=0.6, panning = False, zooming = False,
             vertical_xlabel=True)
        
        #pandas_bokeh.output_file(dirName + "ALL_pyramid.html")
        centres = all_pyramid.T.columns.tolist()
        index = all_pyramid.T.index.tolist()
        pyramid = {}
        xmax = 0
        for gender in ['Men','Women']:
            pyramid[gender]={'index' : all_pyramid.T[all_pyramid.T.index.get_level_values(level=1) == gender].index.tolist()}
            for centre in centres:
                pyramid[gender][centre] = all_pyramid.T[all_pyramid.T.index.get_level_values(level=1) == gender][centre].tolist()
                pyramid[gender]['All'] = all_pyramid.T[all_pyramid.T.index.get_level_values(level=1) == gender].sum(axis=1)
                try:
                    xmax = max(xmax, max(pyramid['Men'][centre]),max(pyramid['Women'][centre]))
                except KeyError:
                    xmax = max(xmax, max(pyramid['Men'][centre]))
        p = figure(y_range = FactorRange(*index) , plot_height=600, plot_width = 800, x_range=(0, 3*xmax), title="Total age/gender distribution",tools="pan,wheel_zoom,box_zoom,reset,hover")
        r1 = p.hbar_stack(centres, y='index', height=0.9, color=GnBu6, source=ColumnDataSource(pyramid['Men']),
                     legend=[" %s" % x for x in centres])
        r2 = p.hbar_stack(centres, y='index', height=0.9, color=OrRd6, source=ColumnDataSource(pyramid['Women']),
                     legend=[" %s" % x for x in centres])
        p.y_range.range_padding = 0.1
        p.ygrid.grid_line_color = None
        p.legend.location = "bottom_right"
        p.xaxis.axis_label = 'Number of cases'
        p.yaxis.axis_label = 'Age group, Gender'   
        p.hover.tooltips = [("Age group, gender", "@index"),
               ("ID","$name"),
               ("Number of cases","@$name"), #name is used for the colors
               ("All centres", "@All")] #"@index $name: @$name"
        #show(p)
        
        pandas_bokeh.output_file(dirName + "All_dashboard.html")
        pandas_bokeh.plot_grid([[p_types]])

        #pandas_bokeh.show(layout)
    return
```

I cannot give you snapshots of data_case and data_demo dataframes as they contain sensitive patient data, but here is a snapshot of the aggregated data that's generated by the code:

data_service

    	week	centre_id	Nosocomial	Comm. acquired	N/A	Type A	Type B	Type N/A	Deaths	No deaths	Medicine	Geriatrics	ICU	Surgery	Paediatrics	Other	All cases
    66	2019-02	Centre #37	0	2	0	2	0	0	0	4	2	0	0	0	0	0	2
    67	2019-02	Centre #38	1	1	0	2	0	0	1	1	1	0	1	0	0	0	2
    68	2019-02	Centre #39	2	4	0	6	0	0	0	9	3	2	0	0	0	1	6
    69	2019-02	Centre #40	8	19	0	27	0	0	0	14	12	6	2	0	2	5	27
    70	2019-02	Centre #43	0	4	0	4	0	0	0	1	3	0	0	0	0	1	4
    71	2019-02	Centre #52	2	5	1	8	0	0	2	8	3	0	0	1	1	3	8

and the pyramid_data:

    centre_id	age_group	Men	Women	other
    0	Centre #37	0-10	0	0	0
    1	Centre #37	10-20	0	1	0
    2	Centre #37	20-30	8	12	0
    3	Centre #37	30-40	5	7	0
    4	Centre #37	40-50	6	7	0
    5	Centre #37	50-60	23	12	0
    6	Centre #37	60-70	24	29	0
    7	Centre #37	70-80	41	46	0
    8	Centre #37	80-90	30	26	0
    9	Centre #37	90-100	5	4	0

Thanks for your help :)

**Comment by PatrikHlobil** [comment]
Hi @PezAmaury , 

sorry for replying so late. I think the reason why your plot fails is related to issue #31 . You are plotting your figures 2 times, first when creating the plot like:
```python
p_types = data_service.groupby('week').sum()[['Type A','Type B','Type N/A']].plot_bokeh.bar(figsize=(800, 600),
             xlabel="Week",
             ylabel="Number of new cases",
             title="Overall distribution of infection type",
             legend="top_right",
             stacked=True,
             alpha=0.6, panning = False, zooming = False,
             vertical_xlabel=True, colormap = colors)
```

and the second time when calling:
```python
pandas_bokeh.plot_grid([[p_types]])
```

Please add to all plot_bokeh calls that generate your figures the keyword parameter `show_figure=False`, like in the example [here](https://github.com/PatrikHlobil/Pandas-Bokeh#dashboard-layouts). The problem only seems to occur when using the output_file option though.

Could this solve your issue?

Best, 

Patrik

**Comment by PezAmaury** [comment]
Hi @PatrikHlobil 

Thanks for the response.
I tried with `show_figure = False` but it initially did not work. With that keyword, I had to do a `save()` so I would save each figure to a HTML file separately (I have to do it as it's a requirement). If I removed the `save()` function, it worked without problems. 

The problem becomes that if I want to save each figure separately AND together, I need to generate them twice: even while using the `plot_grid` function for all of them, then each of them, or follow the `plot_grid` by `save()` functions to save the plots separately, then the runtime error comes again if I don't create the plots a second time. I do not know if this is a pandas_bokeh specific issue or an overall bokeh issue, but this is the only workaround that ended up allowing me to save both a dashbord and each plts separately.

Thank you for your help in this matter!

Best,
Amaury
